Frameworks/Foundation/NSURLConnection.mm

```diff
@@ -475,28 +482,40 @@ - (void)URLProtocol:(NSURLProtocol*)urlProtocol didReceiveAuthenticationChalleng
  @Status Interoperable
 */
 - (void)URLProtocol:(NSURLProtocol*)urlProtocol wasRedirectedToRequest:(NSURLRequest*)request redirectResponse:(NSURLResponse*)response {
+    __block StrongId<NSURLRequest> newRequest = request;
+
+    auto delegateCount = std::make_shared<std::atomic<uint>>(std::max(_scheduledRunLoops.size(), 1u));
+    auto stoppedLoading = std::make_shared<std::atomic<uint>>(0);
+
     __dispatchDelegateCallback(self, ^void() {
-        [_protocol stopLoading];
-        NSURLRequest* newRequest = request;
-        if ([_delegate respondsToSelector:@selector(connection:willSendRequest:redirectResponse:)]) {
-            newRequest = [_delegate connection:self willSendRequest:request redirectResponse:response];
+        // Call stop loading on the delegate thread to preserve callback ordering, only call it once though
+        if (stoppedLoading->fetch_add(1) == 0) {
+            [_protocol stopLoading];
         }
-        _protocol = nil;
 
-        if (!newRequest) {
-            [self cancel];
-            return;
+        if ([_delegate respondsToSelector:@selector(connection:willSendRequest:redirectResponse:)]) {
+            newRequest = [_delegate connection:self willSendRequest:newRequest redirectResponse:response];
         }
 
-        NSURLProtocol* protocol = __protocolForRequest(request, self);
-        if (!protocol) {
-            TraceError(TAG, L"NSURLConnection was redirected to a new request, but could not create a NSURLProtocol for the new request.");
-            return;
-        }
+        if (--*delegateCount == 0) {
+            _protocol = nil;
 
-        _protocol = protocol;
-        _currentRequest = newRequest;
-        [_protocol performSelector:@selector(startLoading) onThread:_protocolThread withObject:nil waitUntilDone:NO];
+            if (!newRequest) {
+                [self cancel];
+                return;
```

Comment:
mm, tried it, doesn't do a good job conveying this being the 'error' case. leaving as-is